### ISSUE TITLE
Issues with Bluetooth in Raspberry Pi

### ISSUE THREAD

**Issue by paulokow** [dbuezas/eq3btsmart#15]
Hi,
The thermostats are not working on my RPI3 setup. Sometimes 3 or 4 connect, but usually none of them work.
The same setup (RPI and thermostats) worked fine with HA prior to 2022.7.

I see the following in my logs:
```
Logger: custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection
Source: custom_components/dbuezas_eq3btsmart/python_eq3bt/eq3bt/bleakconnection.py:124
Integration: dbuezas EQ3 Bluetooth Smart Thermostats
First occurred: 12 listopada 2022 14:08:49 (5725 occurrences)
Last logged: 13:31:53

[Łazienka mała] Broken connection [retry 1/1]: Can't find device
[Salon okno małe] Broken connection [retry 1/1]: Salon okno małe - 00:1A:22:09:B9:D5 -> /org/bluez/hci0: Failed to connect:
[Korytarz] Broken connection [retry 1/1]: Korytarz - 00:1A:22:0B:A1:B3 -> /org/bluez/hci0: Failed to connect:
[Kacper] Broken connection [retry 1/1]: Kacper - 00:1A:22:0B:7B:2D -> /org/bluez/hci0: Failed to connect:
[Salon okno duże] Broken connection [retry 1/1]: Salon okno duże - 00:1A:22:09:94:30 -> /org/bluez/hci0: Failed to connect:
```

And then:
```
Logger: custom_components.dbuezas_eq3btsmart.climate
Source: custom_components/dbuezas_eq3btsmart/climate.py:369
Integration: dbuezas EQ3 Bluetooth Smart Thermostats
First occurred: 12 listopada 2022 14:08:49 (5725 occurrences)
Last logged: 13:31:53

[Łazienka mała] Error updating, will retry later: Can't find device
[Salon okno małe] Error updating, will retry later: Salon okno małe - 00:1A:22:09:B9:D5 -> /org/bluez/hci0: Failed to connect:
[Korytarz] Error updating, will retry later: Korytarz - 00:1A:22:0B:A1:B3 -> /org/bluez/hci0: Failed to connect:
[Kacper] Error updating, will retry later: Kacper - 00:1A:22:0B:7B:2D -> /org/bluez/hci0: Failed to connect:
[Salon okno duże] Error updating, will retry later: Salon okno duże - 00:1A:22:09:94:30 -> /org/bluez/hci0: Failed to connect:
```

Finally
```
Logger: homeassistant.components.bluetooth.scanner
Source: components/bluetooth/scanner.py:364
Integration: Bluetooth ([documentation](https://www.home-assistant.io/integrations/bluetooth), [issues](https://github.com/home-assistant/home-assistant/issues?q=is%3Aissue+is%3Aopen+label%3A%22integration%3A+bluetooth%22))
First occurred: 14 listopada 2022 08:59:37 (57 occurrences)
Last logged: 13:32:46

hci0 (B8:27:EB:53:6F:2B): Error stopping scanner: [org.bluez.Error.InProgress] Operation already in progress
hci0 (B8:27:EB:53:6F:2B): Error stopping scanner: [org.bluez.Error.NotReady] Resource Not Ready
```

```
Logger: bluetooth_auto_recovery.recover
Source: components/bluetooth/util.py:112
First occurred: 14 listopada 2022 09:01:40 (47 occurrences)
Last logged: 13:32:56

Bluetooth adapter hci0 could not be reset due to timeout
Bluetooth adapter hci0 successfully turned back ON
```

and

```
Logger: homeassistant.components.bluetooth.scanner
Source: components/bluetooth/scanner.py:322
Integration: Bluetooth ([documentation](https://www.home-assistant.io/integrations/bluetooth), [issues](https://github.com/home-assistant/home-assistant/issues?q=is%3Aissue+is%3Aopen+label%3A%22integration%3A+bluetooth%22))
First occurred: 15 listopada 2022 04:42:19 (9 occurrences)
Last logged: 13:32:56

hci0 (B8:27:EB:53:6F:2B): Failed to restart Bluetooth scanner: hci0 (B8:27:EB:53:6F:2B): Failed to start Bluetooth: [org.bluez.Error.NotReady] Resource Not Ready
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/bluetooth/scanner.py", line 241, in _async_start
    await self.scanner.start()  # type: ignore[no-untyped-call]
  File "/usr/local/lib/python3.10/site-packages/bleak/__init__.py", line 156, in start
    await self._backend.start()
  File "/usr/local/lib/python3.10/site-packages/bleak/backends/bluezdbus/scanner.py", line 191, in start
    self._stop = await manager.active_scan(
  File "/usr/local/lib/python3.10/site-packages/bleak/backends/bluezdbus/manager.py", line 357, in active_scan
    assert_reply(reply)
  File "/usr/local/lib/python3.10/site-packages/bleak/backends/bluezdbus/utils.py", line 20, in assert_reply
    raise BleakDBusError(reply.error_name, reply.body)
bleak.exc.BleakDBusError: [org.bluez.Error.NotReady] Resource Not Ready

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/bluetooth/scanner.py", line 374, in _async_scanner_watchdog
    await self._async_start()
  File "/usr/src/homeassistant/homeassistant/components/bluetooth/scanner.py", line 322, in _async_start
    raise ScannerStartError(
homeassistant.components.bluetooth.scanner.ScannerStartError: hci0 (B8:27:EB:53:6F:2B): Failed to start Bluetooth: [org.bluez.Error.NotReady] Resource Not Ready
```

**Comment by dbuezas** [comment]
There seems to be an issue with the bluetooth dongle. Can you connect to the thermostats with bluetoothctl? (see pairing section of readme)

I had that once and I had to physically power cycle the dongle for it to start working again (unplug it from the usb).

**Comment by paulokow** [comment]
@dbuezas I will give it a go. Just as clarification, it's not a dongle, the BT is the built-in one in RPI3.

**Comment by paulokow** [comment]
This is a short output from `bluetoothctl`. I can see info of one of the devices, which do not update in HA.
```
[bluetooth]# info 00:1A:22:09:94:30
Device 00:1A:22:09:94:30 (public)
	Name: CC-RT-BLE
	Alias: CC-RT-BLE
	Paired: no
	Trusted: yes
	Blocked: no
	Connected: no
	LegacyPairing: no
	UUID: Generic Access Profile    (00001800-0000-1000-8000-00805f9b34fb)
	UUID: Generic Attribute Profile (00001801-0000-1000-8000-00805f9b34fb)
	UUID: Device Information        (0000180a-0000-1000-8000-00805f9b34fb)
	ManufacturerData Key: 0x0000
	ManufacturerData Value:
  00 00 00 00 00 00 00 00 00                       .........       
	RSSI: -81
[CHG] Device FE:9F:67:F0:B1:7E RSSI: -84
[CHG] Device 00:1A:22:0B:7E:39 RSSI: -86
[CHG] Device 7C:63:16:07:0C:59 RSSI: -88
[CHG] Device 00:1A:22:0B:7B:2D RSSI: -87
[CHG] Device 00:1A:22:0B:6C:4B RSSI: -87
[CHG] Device 45:B9:9B:C8:E3:97 RSSI: -93
[CHG] Device E8:BC:07:85:A6:1F RSSI: -80
[CHG] Device C5:8C:9B:BF:B4:51 RSSI: -78
[bluetooth]# devices
Device 7C:63:16:07:0C:59 7C-63-16-07-0C-59
Device 7E:75:B0:40:B7:25 7E-75-B0-40-B7-25
Device 0E:2D:0F:1C:31:00 0E-2D-0F-1C-31-00
Device FE:9F:67:F0:B1:7E FE-9F-67-F0-B1-7E
Device 00:1A:22:0B:6C:A3 CC-RT-BLE
Device E8:BC:07:85:A6:1F E8-BC-07-85-A6-1F
Device C5:8C:9B:BF:B4:51 C5-8C-9B-BF-B4-51
Device 69:41:7A:C8:56:1E 69-41-7A-C8-56-1E
Device 45:73:5A:F6:F0:E1 45-73-5A-F6-F0-E1
Device 61:45:C8:43:1A:55 61-45-C8-43-1A-55
Device 45:B9:9B:C8:E3:97 45-B9-9B-C8-E3-97
Device 77:E0:4B:32:70:D9 77-E0-4B-32-70-D9
Device 68:0C:61:9D:82:99 68-0C-61-9D-82-99
Device 38:68:A4:75:21:EB 38-68-A4-75-21-EB
Device 00:1A:22:0B:7E:39 CC-RT-BLE
Device 28:11:A5:B9:CD:C9 LE-Bose Revolve SoundLink
Device 00:1A:22:09:B9:61 CC-RT-BLE
Device 00:1A:22:0B:6C:4B CC-RT-BLE
Device 10:2B:41:75:C8:AD 10-2B-41-75-C8-AD
Device B0:22:7A:9C:C8:8B B0-22-7A-9C-C8-8B
Device 2C:4C:C6:0B:BB:95 Venus_2C4CC60BBB95
Device 00:1A:22:09:B9:D5 CC-RT-BLE
Device 00:1A:22:0B:7B:2D CC-RT-BLE
Device 00:1A:22:09:94:30 CC-RT-BLE
Device 00:1A:22:0B:A1:B3 CC-RT-BLE
Device 3B:81:98:23:61:6E 3B-81-98-23-61-6E
Device 46:7B:A8:E9:3A:4F 46-7B-A8-E9-3A-4F
Device 16:5A:D6:A0:9C:ED 16-5A-D6-A0-9C-ED
Device 5E:8B:01:27:18:E7 5E-8B-01-27-18-E7
```

**Comment by paulokow** [comment]
OK, now pairing is also done, but no difference. Just a note, pairing went through without any pin prompt
```
[bluetooth]# info 00:1A:22:09:94:30
Device 00:1A:22:09:94:30 (public)
	Name: CC-RT-BLE
	Alias: CC-RT-BLE
	Paired: yes
	Trusted: yes
	Blocked: no
	Connected: no
	LegacyPairing: no
	UUID: Generic Access Profile    (00001800-0000-1000-8000-00805f9b34fb)
	UUID: Generic Attribute Profile (00001801-0000-1000-8000-00805f9b34fb)
	UUID: Device Information        (0000180a-0000-1000-8000-00805f9b34fb)
	ManufacturerData Key: 0x0000
	ManufacturerData Value:
  00 00 00 00 00 00 00 00 00                       .........       
	RSSI: -81
```

**Comment by paulokow** [comment]
I tried to enable debug logs, but somehow no other output is visible
```
logger:
  default: warning
  logs:
    custom_components.dbuezas_eq3btsmart.climate: debug
```

**Comment by dbuezas** [comment]
What happens if you connect to the thermostat with bluetoothctl?

**Comment by paulokow** [comment]
It shows connected ok after a short while.

I now managed to get the logs (the logs goes from the fresh reboot):
```
2022-11-17 10:42:36.617 WARNING (SyncWorker_1) [homeassistant.loader] We found a custom integration dbuezas_eq3btsmart which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2022-11-17 10:43:18.018 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Querying the device..
2022-11-17 10:43:18.019 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno małe] Connecting with ble_device, rssi: -77
2022-11-17 10:43:18.205 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Korytarz] Querying the device..
2022-11-17 10:43:18.208 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Korytarz] Connecting with ble_device, rssi: -80
2022-11-17 10:43:18.227 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Kacper] Querying the device..
2022-11-17 10:43:18.229 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Kacper] Connecting with ble_device, rssi: -81
2022-11-17 10:43:18.231 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno duże] Querying the device..
2022-11-17 10:43:18.251 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno duże] Connecting with ble_device, rssi: -82
2022-11-17 10:43:53.502 WARNING (MainThread) [homeassistant.components.climate] Setup of climate platform dbuezas_eq3btsmart is taking over 10 seconds.
2022-11-17 10:43:53.506 WARNING (MainThread) [homeassistant.components.climate] Setup of climate platform dbuezas_eq3btsmart is taking over 10 seconds.
2022-11-17 10:43:53.509 WARNING (MainThread) [homeassistant.components.climate] Setup of climate platform dbuezas_eq3btsmart is taking over 10 seconds.
2022-11-17 10:43:53.513 WARNING (MainThread) [homeassistant.components.climate] Setup of climate platform dbuezas_eq3btsmart is taking over 10 seconds.
2022-11-17 10:43:57.853 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] Salon okno małe: Disconnected from device; rssi: -77
2022-11-17 10:44:01.719 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno małe] Connected
2022-11-17 10:44:01.739 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno małe] Paired: True 
2022-11-17 10:44:02.168 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Received notification from the device.
2022-11-17 10:44:02.178 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Got status: b'020109000409'
2022-11-17 10:44:02.181 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Parsed status: Container: 
2022-11-17 10:44:02.182 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Valve state:      0
2022-11-17 10:44:02.182 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Mode:             manual off dst
2022-11-17 10:44:02.183 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Target temp:      4.5
2022-11-17 10:44:02.183 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Away end:         None
2022-11-17 10:44:02.183 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Window open temp: None
2022-11-17 10:44:02.183 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Window open time: None
2022-11-17 10:44:02.184 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Comfort temp:     None
2022-11-17 10:44:02.184 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Eco temp:         None
2022-11-17 10:44:02.186 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Temp offset:      None
2022-11-17 10:44:05.724 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Kacper] Connected
2022-11-17 10:44:05.758 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Kacper] Paired: True 
2022-11-17 10:44:05.759 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Kacper] Broken connection [retry 1/1]: Characteristic d0e8434d-cd29-0996-af41-6c90f4e0eb2a not found!
2022-11-17 10:44:05.773 ERROR (MainThread) [custom_components.dbuezas_eq3btsmart.climate] [Kacper] Error updating, will retry later: Characteristic d0e8434d-cd29-0996-af41-6c90f4e0eb2a not found!
2022-11-17 10:44:18.006 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno duże] Connected
2022-11-17 10:44:18.029 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno duże] Paired: True 
2022-11-17 10:44:18.030 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno duże] Broken connection [retry 1/1]: Characteristic d0e8434d-cd29-0996-af41-6c90f4e0eb2a not found!
2022-11-17 10:44:18.037 ERROR (MainThread) [custom_components.dbuezas_eq3btsmart.climate] [Salon okno duże] Error updating, will retry later: Characteristic d0e8434d-cd29-0996-af41-6c90f4e0eb2a not found!
2022-11-17 10:44:29.978 WARNING (MainThread) [homeassistant.bootstrap] Waiting on integrations to complete setup: climate.dbuezas_eq3btsmart_2
2022-11-17 10:44:32.204 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Korytarz] Connected
2022-11-17 10:44:32.252 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Korytarz] Paired: True 
2022-11-17 10:44:32.253 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Korytarz] Broken connection [retry 1/1]: Characteristic d0e8434d-cd29-0996-af41-6c90f4e0eb2a not found!
2022-11-17 10:44:32.260 ERROR (MainThread) [custom_components.dbuezas_eq3btsmart.climate] [Korytarz] Error updating, will retry later: Characteristic d0e8434d-cd29-0996-af41-6c90f4e0eb2a not found!
2022-11-17 10:44:39.981 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.config_flow] Discovered eQ3 thermostat using bluetooth: BluetoothServiceInfoBleak(name='CC-RT-BLE', address='00:1A:22:09:94:30', rssi=-82, manufacturer_data={0: b'\x00\x00\x00\x00\x00\x00\x00\x00\x00'}, service_data={}, service_uuids=['00001800-0000-1000-8000-00805f9b34fb', '00001801-0000-1000-8000-00805f9b34fb', '0000180a-0000-1000-8000-00805f9b34fb'], source='B8:27:EB:53:6F:2B', device=BLEDevice(00:1A:22:09:94:30, CC-RT-BLE), advertisement=AdvertisementData(local_name='CC-RT-BLE', manufacturer_data={0: b'\x00\x00\x00\x00\x00\x00\x00\x00\x00'}, service_uuids=['00001800-0000-1000-8000-00805f9b34fb', '00001801-0000-1000-8000-00805f9b34fb', '0000180a-0000-1000-8000-00805f9b34fb'], rssi=-82), connectable=True, time=116.207238101), CC-RT-BLE
2022-11-17 10:44:40.057 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.config_flow] Discovered eQ3 thermostat using bluetooth: BluetoothServiceInfoBleak(name='CC-RT-BLE', address='00:1A:22:0B:A1:B3', rssi=-80, manufacturer_data={0: b'\x00\x00\x00\x00\x00\x00\x00\x00\x00'}, service_data={}, service_uuids=['00001800-0000-1000-8000-00805f9b34fb', '00001801-0000-1000-8000-00805f9b34fb', '0000180a-0000-1000-8000-00805f9b34fb', '9e5d1e47-5c13-43a0-8635-82ad38a1386f'], source='B8:27:EB:53:6F:2B', device=BLEDevice(00:1A:22:0B:A1:B3, CC-RT-BLE), advertisement=AdvertisementData(local_name='CC-RT-BLE', manufacturer_data={0: b'\x00\x00\x00\x00\x00\x00\x00\x00\x00'}, service_uuids=['00001800-0000-1000-8000-00805f9b34fb', '00001801-0000-1000-8000-00805f9b34fb', '0000180a-0000-1000-8000-00805f9b34fb', '9e5d1e47-5c13-43a0-8635-82ad38a1386f'], rssi=-80), connectable=True, time=116.443254725), CC-RT-BLE
2022-11-17 10:44:40.924 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.config_flow] Discovered eQ3 thermostat using bluetooth: BluetoothServiceInfoBleak(name='CC-RT-BLE', address='00:1A:22:09:B9:D5', rssi=-77, manufacturer_data={0: b'\x00\x00\x00\x00\x00\x00\x00\x00\x00'}, service_data={}, service_uuids=['00001800-0000-1000-8000-00805f9b34fb', '00001801-0000-1000-8000-00805f9b34fb', '0000180a-0000-1000-8000-00805f9b34fb', '3e135142-654f-9090-134a-a6ff5bb77046', '9e5d1e47-5c13-43a0-8635-82ad38a1386f'], source='B8:27:EB:53:6F:2B', device=BLEDevice(00:1A:22:09:B9:D5, CC-RT-BLE), advertisement=AdvertisementData(local_name='CC-RT-BLE', manufacturer_data={0: b'\x00\x00\x00\x00\x00\x00\x00\x00\x00'}, service_uuids=['00001800-0000-1000-8000-00805f9b34fb', '00001801-0000-1000-8000-00805f9b34fb', '0000180a-0000-1000-8000-00805f9b34fb', '3e135142-654f-9090-134a-a6ff5bb77046', '9e5d1e47-5c13-43a0-8635-82ad38a1386f'], rssi=-77), connectable=True, time=116.732275083), CC-RT-BLE
2022-11-17 10:44:40.971 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.config_flow] Discovered eQ3 thermostat using bluetooth: BluetoothServiceInfoBleak(name='CC-RT-BLE', address='00:1A:22:0B:6C:4B', rssi=-88, manufacturer_data={0: b'\x00\x00\x00\x00\x00\x00\x00\x00\x00'}, service_data={}, service_uuids=['3e135142-654f-9090-134a-a6ff5bb77046'], source='B8:27:EB:53:6F:2B', device=BLEDevice(00:1A:22:0B:6C:4B, CC-RT-BLE), advertisement=AdvertisementData(local_name='CC-RT-BLE', manufacturer_data={0: b'\x00\x00\x00\x00\x00\x00\x00\x00\x00'}, service_uuids=['3e135142-654f-9090-134a-a6ff5bb77046'], rssi=-88), connectable=True, time=116.733275153), CC-RT-BLE
2022-11-17 10:44:40.973 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.config_flow] Discovered eQ3 thermostat using bluetooth: BluetoothServiceInfoBleak(name='CC-RT-BLE', address='00:1A:22:0B:7B:2D', rssi=-81, manufacturer_data={0: b'\x00\x00\x00\x00\x00\x00\x00\x00\x00'}, service_data={}, service_uuids=['00001800-0000-1000-8000-00805f9b34fb', '00001801-0000-1000-8000-00805f9b34fb', '9e5d1e47-5c13-43a0-8635-82ad38a1386f'], source='B8:27:EB:53:6F:2B', device=BLEDevice(00:1A:22:0B:7B:2D, CC-RT-BLE), advertisement=AdvertisementData(local_name='CC-RT-BLE', manufacturer_data={0: b'\x00\x00\x00\x00\x00\x00\x00\x00\x00'}, service_uuids=['00001800-0000-1000-8000-00805f9b34fb', '00001801-0000-1000-8000-00805f9b34fb', '9e5d1e47-5c13-43a0-8635-82ad38a1386f'], rssi=-81), connectable=True, time=116.79427945), CC-RT-BLE
2022-11-17 10:44:41.018 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.config_flow] Discovered eQ3 thermostat using bluetooth: BluetoothServiceInfoBleak(name='CC-RT-BLE', address='00:1A:22:0B:7E:39', rssi=-84, manufacturer_data={0: b'\x00\x00\x00\x00\x00\x00\x00\x00\x00'}, service_data={}, service_uuids=['3e135142-654f-9090-134a-a6ff5bb77046'], source='B8:27:EB:53:6F:2B', device=BLEDevice(00:1A:22:0B:7E:39, CC-RT-BLE), advertisement=AdvertisementData(local_name='CC-RT-BLE', manufacturer_data={0: b'\x00\x00\x00\x00\x00\x00\x00\x00\x00'}, service_uuids=['3e135142-654f-9090-134a-a6ff5bb77046'], rssi=-84), connectable=True, time=117.486327216), CC-RT-BLE
2022-11-17 10:44:42.774 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.config_flow] Discovered eQ3 thermostat using bluetooth: BluetoothServiceInfoBleak(name='CC-RT-BLE', address='00:1A:22:09:B9:61', rssi=-81, manufacturer_data={}, service_data={}, service_uuids=['3e135142-654f-9090-134a-a6ff5bb77046'], source='B8:27:EB:53:6F:2B', device=BLEDevice(00:1A:22:09:B9:61, CC-RT-BLE), advertisement=AdvertisementData(local_name='CC-RT-BLE', service_uuids=['3e135142-654f-9090-134a-a6ff5bb77046'], rssi=-81), connectable=True, time=124.17275008), CC-RT-BLE
2022-11-17 10:44:42.776 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.config_flow] Discovered eQ3 thermostat using bluetooth: BluetoothServiceInfoBleak(name='CC-RT-BLE', address='00:1A:22:09:B9:61', rssi=-81, manufacturer_data={0: b'\x00\x00\x00\x00\x00\x00\x00\x00\x00'}, service_data={}, service_uuids=['3e135142-654f-9090-134a-a6ff5bb77046'], source='B8:27:EB:53:6F:2B', device=BLEDevice(00:1A:22:09:B9:61, CC-RT-BLE), advertisement=AdvertisementData(local_name='CC-RT-BLE', manufacturer_data={0: b'\x00\x00\x00\x00\x00\x00\x00\x00\x00'}, service_uuids=['3e135142-654f-9090-134a-a6ff5bb77046'], rssi=-94), connectable=True, time=129.244028508), CC-RT-BLE
2022-11-17 10:44:42.795 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.config_flow] Discovered eQ3 thermostat using bluetooth: BluetoothServiceInfoBleak(name='CC-RT-BLE', address='00:1A:22:0B:6C:A3', rssi=-97, manufacturer_data={}, service_data={}, service_uuids=['3e135142-654f-9090-134a-a6ff5bb77046'], source='B8:27:EB:53:6F:2B', device=BLEDevice(00:1A:22:0B:6C:A3, CC-RT-BLE), advertisement=AdvertisementData(local_name='CC-RT-BLE', service_uuids=['3e135142-654f-9090-134a-a6ff5bb77046'], rssi=-97), connectable=True, time=172.393529338), CC-RT-BLE
2022-11-17 10:44:42.797 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.config_flow] Discovered eQ3 thermostat using bluetooth: BluetoothServiceInfoBleak(name='CC-RT-BLE', address='00:1A:22:0B:6C:A3', rssi=-97, manufacturer_data={0: b'\x00\x00\x00\x00\x00\x00\x00\x00\x00'}, service_data={}, service_uuids=['3e135142-654f-9090-134a-a6ff5bb77046'], source='B8:27:EB:53:6F:2B', device=BLEDevice(00:1A:22:0B:6C:A3, CC-RT-BLE), advertisement=AdvertisementData(local_name='CC-RT-BLE', manufacturer_data={0: b'\x00\x00\x00\x00\x00\x00\x00\x00\x00'}, service_uuids=['3e135142-654f-9090-134a-a6ff5bb77046'], rssi=-95), connectable=True, time=177.819655255), CC-RT-BLE
2022-11-17 10:45:33.772 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] Salon okno małe: Disconnected from device; rssi: -77
2022-11-17 10:45:37.423 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] Kacper: Disconnected from device; rssi: -81
2022-11-17 10:45:49.850 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] Salon okno duże: Disconnected from device; rssi: -82
2022-11-17 10:46:04.097 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] Korytarz: Disconnected from device; rssi: -80
2022-11-17 10:49:02.206 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Querying the device..
2022-11-17 10:49:02.207 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno małe] Connecting with ble_device, rssi: -77
2022-11-17 10:49:05.784 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Kacper] Querying the device..
2022-11-17 10:49:05.785 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Kacper] Connecting with ble_device, rssi: -81
2022-11-17 10:49:09.449 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno małe] Connected
2022-11-17 10:49:09.464 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno małe] Paired: True 
2022-11-17 10:49:09.901 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Received notification from the device.
2022-11-17 10:49:09.901 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Got status: b'020109000409'
2022-11-17 10:49:09.905 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Parsed status: Container: 
2022-11-17 10:49:09.906 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Valve state:      0
2022-11-17 10:49:09.909 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Mode:             manual off dst
2022-11-17 10:49:09.910 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Target temp:      4.5
2022-11-17 10:49:09.910 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Away end:         None
2022-11-17 10:49:09.911 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Window open temp: None
2022-11-17 10:49:09.911 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Window open time: None
2022-11-17 10:49:09.911 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Comfort temp:     None
2022-11-17 10:49:09.911 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Eco temp:         None
2022-11-17 10:49:09.912 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Temp offset:      None
2022-11-17 10:49:18.049 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno duże] Querying the device..
2022-11-17 10:49:18.050 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno duże] Connecting with ble_device, rssi: -82
2022-11-17 10:49:22.679 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno duże] Connected
2022-11-17 10:49:22.697 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno duże] Paired: True 
2022-11-17 10:49:22.698 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno duże] Broken connection [retry 1/1]: Characteristic d0e8434d-cd29-0996-af41-6c90f4e0eb2a not found!
2022-11-17 10:49:22.704 ERROR (MainThread) [custom_components.dbuezas_eq3btsmart.climate] [Salon okno duże] Error updating, will retry later: Characteristic d0e8434d-cd29-0996-af41-6c90f4e0eb2a not found!
2022-11-17 10:49:32.284 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Korytarz] Querying the device..
2022-11-17 10:49:32.285 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Korytarz] Connecting with ble_device, rssi: -80
2022-11-17 10:49:52.080 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Korytarz] Connected
2022-11-17 10:49:52.095 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Korytarz] Paired: True 
2022-11-17 10:49:52.096 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Korytarz] Broken connection [retry 1/1]: Characteristic d0e8434d-cd29-0996-af41-6c90f4e0eb2a not found!
2022-11-17 10:49:52.103 ERROR (MainThread) [custom_components.dbuezas_eq3btsmart.climate] [Korytarz] Error updating, will retry later: Characteristic d0e8434d-cd29-0996-af41-6c90f4e0eb2a not found!
2022-11-17 10:50:02.370 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Kacper] Connected
2022-11-17 10:50:02.387 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Kacper] Paired: True 
2022-11-17 10:50:02.387 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Kacper] Broken connection [retry 1/1]: Characteristic d0e8434d-cd29-0996-af41-6c90f4e0eb2a not found!
2022-11-17 10:50:02.394 ERROR (MainThread) [custom_components.dbuezas_eq3btsmart.climate] [Kacper] Error updating, will retry later: Characteristic d0e8434d-cd29-0996-af41-6c90f4e0eb2a not found!
2022-11-17 10:50:41.408 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] Salon okno małe: Disconnected from device; rssi: -77
2022-11-17 10:50:54.476 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] Salon okno duże: Disconnected from device; rssi: -82
2022-11-17 10:51:23.780 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] Korytarz: Disconnected from device; rssi: -80
2022-11-17 10:51:34.167 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] Kacper: Disconnected from device; rssi: -81
2022-11-17 10:53:05.984 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.config_flow] Discovered eQ3 thermostat using bluetooth: BluetoothServiceInfoBleak(name='CC-RT-BLE', address='00:1A:22:0B:6C:4B', rssi=-88, manufacturer_data={0: b'\x00\x00\x00\x00\x00\x00\x00\x00\x00'}, service_data={}, service_uuids=['3e135142-654f-9090-134a-a6ff5bb77046', '3e000442-654f-9090-134a-a6ff5bb77046'], source='B8:27:EB:53:6F:2B', device=BLEDevice(00:1A:22:0B:6C:4B, CC-RT-BLE), advertisement=AdvertisementData(local_name='CC-RT-BLE', manufacturer_data={0: b'\x00\x00\x00\x00\x00\x00\x00\x00\x00'}, service_uuids=['3e135142-654f-9090-134a-a6ff5bb77046', '3e000442-654f-9090-134a-a6ff5bb77046'], rssi=36), connectable=True, time=724.906076346), CC-RT-BLE
2022-11-17 10:54:02.208 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Querying the device..
2022-11-17 10:54:02.209 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno małe] Connecting with ble_device, rssi: -77
2022-11-17 10:54:05.786 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Kacper] Querying the device..
2022-11-17 10:54:05.787 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Kacper] Connecting with ble_device, rssi: -81
2022-11-17 10:54:13.402 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Kacper] Connected
2022-11-17 10:54:13.423 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Kacper] Paired: True 
2022-11-17 10:54:13.424 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Kacper] Broken connection [retry 1/1]: Characteristic d0e8434d-cd29-0996-af41-6c90f4e0eb2a not found!
2022-11-17 10:54:13.432 ERROR (MainThread) [custom_components.dbuezas_eq3btsmart.climate] [Kacper] Error updating, will retry later: Characteristic d0e8434d-cd29-0996-af41-6c90f4e0eb2a not found!
2022-11-17 10:54:14.220 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno małe] Connected
2022-11-17 10:54:14.235 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno małe] Paired: True 
2022-11-17 10:54:14.672 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Received notification from the device.
2022-11-17 10:54:14.673 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Got status: b'020109000409'
2022-11-17 10:54:14.675 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Parsed status: Container: 
2022-11-17 10:54:14.677 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Valve state:      0
2022-11-17 10:54:14.679 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Mode:             manual off dst
2022-11-17 10:54:14.680 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Target temp:      4.5
2022-11-17 10:54:14.680 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Away end:         None
2022-11-17 10:54:14.681 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Window open temp: None
2022-11-17 10:54:14.681 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Window open time: None
2022-11-17 10:54:14.682 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Comfort temp:     None
2022-11-17 10:54:14.682 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Eco temp:         None
2022-11-17 10:54:14.682 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Temp offset:      None
2022-11-17 10:54:18.054 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno duże] Querying the device..
2022-11-17 10:54:18.055 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno duże] Connecting with ble_device, rssi: -82
2022-11-17 10:54:28.425 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno duże] Connected
2022-11-17 10:54:28.445 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno duże] Paired: True 
2022-11-17 10:54:28.446 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno duże] Broken connection [retry 1/1]: Characteristic d0e8434d-cd29-0996-af41-6c90f4e0eb2a not found!
2022-11-17 10:54:28.452 ERROR (MainThread) [custom_components.dbuezas_eq3btsmart.climate] [Salon okno duże] Error updating, will retry later: Characteristic d0e8434d-cd29-0996-af41-6c90f4e0eb2a not found!
```

**Comment by dbuezas** [comment]
It seems like the thermostats don't accept the request for notifications. This is really odd. 

I can try making a version that uses handles (a short number) instead of uids (more specific long list of alphanumeric values), but first make sure you are on the latest HA version and turn the pi filly off (also take remove power) to ensure its not some random bug somewhere else.

Let me know the result and I'll try to make a test version after that.

**Comment by dbuezas** [comment]
I had some time and made the test version. In HACS, go to this component and click "show beta versions".
But first please try what I ask in the previous comment.

**Comment by dbuezas** [comment]
you have to click this component, then "redownload" under the 3 vertical dots and toggle "show beta"
<img width="545" alt="image" src="https://user-images.githubusercontent.com/777196/202576902-b784cc2f-8905-4778-9ca6-f49b88e4f4ee.png">

**Comment by paulokow** [comment]
@dbuezas thanks for the effort you are putting into this issue.
Power-cycle did not help - neither with 1.0.3 nor with the experimental version.

The logs from the experimental version. Here switch to Auto mode.

Question - can you tell what commands I can issue through bluetoothctl to simulate the same thing that the driver is doing? Just to make sure that there is no other issue related to h/w or connection. Interesting enough one of the thermostats sometimes works, just the other 3 not. And it's always the same one.

```
2022-11-18 11:36:23.219 ERROR (MainThread) [custom_components.dbuezas_eq3btsmart.climate] [Korytarz] Error updating, will retry later: Can't find device
2022-11-18 11:36:24.920 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Kacper] Querying the device..
2022-11-18 11:36:24.921 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Kacper] Broken connection [retry 1/1]: Can't find device
2022-11-18 11:36:24.927 ERROR (MainThread) [custom_components.dbuezas_eq3btsmart.climate] [Kacper] Error updating, will retry later: Can't find device
2022-11-18 11:36:27.890 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno duże] Querying the device..
2022-11-18 11:36:27.890 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno duże] Broken connection [retry 1/1]: Can't find device
2022-11-18 11:36:27.897 ERROR (MainThread) [custom_components.dbuezas_eq3btsmart.climate] [Salon okno duże] Error updating, will retry later: Can't find device
2022-11-18 11:36:30.911 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno małe] Querying the device..
2022-11-18 11:36:30.912 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno małe] Broken connection [retry 1/1]: Can't find device
2022-11-18 11:36:30.920 ERROR (MainThread) [custom_components.dbuezas_eq3btsmart.climate] [Salon okno małe] Error updating, will retry later: Can't find device
2022-11-18 11:38:45.962 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Salon okno duże] Setting new mode: Mode.Auto
2022-11-18 11:38:45.962 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno duże] Broken connection [retry 1/10]: Can't find device
2022-11-18 11:38:46.980 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno duże] Broken connection [retry 2/10]: Can't find device
2022-11-18 11:38:47.989 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno duże] Broken connection [retry 3/10]: Can't find device
2022-11-18 11:38:48.135 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Korytarz] Setting new mode: Mode.Auto
2022-11-18 11:38:48.135 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Korytarz] Broken connection [retry 1/10]: Can't find device
2022-11-18 11:38:48.998 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno duże] Broken connection [retry 4/10]: Can't find device
2022-11-18 11:38:49.153 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Korytarz] Broken connection [retry 2/10]: Can't find device
2022-11-18 11:38:50.006 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno duże] Broken connection [retry 5/10]: Can't find device
2022-11-18 11:38:50.161 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Korytarz] Broken connection [retry 3/10]: Can't find device
2022-11-18 11:38:51.015 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno duże] Broken connection [retry 6/10]: Can't find device
2022-11-18 11:38:51.168 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Korytarz] Broken connection [retry 4/10]: Can't find device
2022-11-18 11:38:51.935 DEBUG (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.eq3btsmart] [Kacper] Setting new mode: Mode.Auto
2022-11-18 11:38:51.936 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Kacper] Broken connection [retry 1/10]: Can't find device
2022-11-18 11:38:52.020 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno duże] Broken connection [retry 7/10]: Can't find device
2022-11-18 11:38:52.176 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Korytarz] Broken connection [retry 5/10]: Can't find device
2022-11-18 11:38:52.969 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Kacper] Broken connection [retry 2/10]: Can't find device
2022-11-18 11:38:53.027 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno duże] Broken connection [retry 8/10]: Can't find device
2022-11-18 11:38:53.182 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Korytarz] Broken connection [retry 6/10]: Can't find device
2022-11-18 11:38:53.975 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Kacper] Broken connection [retry 3/10]: Can't find device
2022-11-18 11:38:54.040 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno duże] Broken connection [retry 9/10]: Can't find device
2022-11-18 11:38:54.193 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Korytarz] Broken connection [retry 7/10]: Can't find device
2022-11-18 11:38:54.994 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Kacper] Broken connection [retry 4/10]: Can't find device
2022-11-18 11:38:55.052 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Salon okno duże] Broken connection [retry 10/10]: Can't find device
2022-11-18 11:38:55.060 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [547907073744] Can't find device
Traceback (most recent call last):
File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 200, in handle_call_service
await hass.services.async_call(
File "/usr/src/homeassistant/homeassistant/core.py", line 1744, in async_call
task.result()
File "/usr/src/homeassistant/homeassistant/core.py", line 1781, in _execute_service
await cast(Callable[[ServiceCall], Awaitable[None]], handler.job.target)(
File "/usr/src/homeassistant/homeassistant/helpers/entity_component.py", line 208, in handle_service
await service.entity_service_call(
File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 678, in entity_service_call
future.result() # pop exception if have
File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 943, in async_request_call
await coro
File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 715, in _handle_entity_call
await result
File "/config/custom_components/dbuezas_eq3btsmart/climate.py", line 245, in async_set_hvac_mode
await self._thermostat.async_set_mode(HA_TO_EQ_HVAC[hvac_mode])
File "/config/custom_components/dbuezas_eq3btsmart/python_eq3bt/eq3bt/eq3btsmart.py", line 315, in async_set_mode
return await self._async_set_mode(0)
File "/config/custom_components/dbuezas_eq3btsmart/python_eq3bt/eq3bt/eq3btsmart.py", line 340, in _async_set_mode
await self._conn.async_make_request(value)
File "/config/custom_components/dbuezas_eq3btsmart/python_eq3bt/eq3bt/bleakconnection.py", line 129, in async_make_request
raise ex
File "/config/custom_components/dbuezas_eq3btsmart/python_eq3bt/eq3bt/bleakconnection.py", line 111, in async_make_request
conn = await self.async_get_connection()
File "/config/custom_components/dbuezas_eq3btsmart/python_eq3bt/eq3bt/bleakconnection.py", line 62, in async_get_connection
raise BackendException("Can't find device")
custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.BackendException: Can't find device
2022-11-18 11:38:55.199 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Korytarz] Broken connection [retry 8/10]: Can't find device
2022-11-18 11:38:56.003 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Kacper] Broken connection [retry 5/10]: Can't find device
2022-11-18 11:38:56.207 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Korytarz] Broken connection [retry 9/10]: Can't find device
2022-11-18 11:38:57.009 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Kacper] Broken connection [retry 6/10]: Can't find device
2022-11-18 11:38:57.224 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Korytarz] Broken connection [retry 10/10]: Can't find device
2022-11-18 11:38:57.230 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [547907073744] Can't find device
Traceback (most recent call last):
File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 200, in handle_call_service
await hass.services.async_call(
File "/usr/src/homeassistant/homeassistant/core.py", line 1744, in async_call
task.result()
File "/usr/src/homeassistant/homeassistant/core.py", line 1781, in _execute_service
await cast(Callable[[ServiceCall], Awaitable[None]], handler.job.target)(
File "/usr/src/homeassistant/homeassistant/helpers/entity_component.py", line 208, in handle_service
await service.entity_service_call(
File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 678, in entity_service_call
future.result() # pop exception if have
File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 943, in async_request_call
await coro
File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 715, in _handle_entity_call
await result
File "/config/custom_components/dbuezas_eq3btsmart/climate.py", line 245, in async_set_hvac_mode
await self._thermostat.async_set_mode(HA_TO_EQ_HVAC[hvac_mode])
File "/config/custom_components/dbuezas_eq3btsmart/python_eq3bt/eq3bt/eq3btsmart.py", line 315, in async_set_mode
return await self._async_set_mode(0)
File "/config/custom_components/dbuezas_eq3btsmart/python_eq3bt/eq3bt/eq3btsmart.py", line 340, in _async_set_mode
await self._conn.async_make_request(value)
File "/config/custom_components/dbuezas_eq3btsmart/python_eq3bt/eq3bt/bleakconnection.py", line 129, in async_make_request
raise ex
File "/config/custom_components/dbuezas_eq3btsmart/python_eq3bt/eq3bt/bleakconnection.py", line 111, in async_make_request
conn = await self.async_get_connection()
File "/config/custom_components/dbuezas_eq3btsmart/python_eq3bt/eq3bt/bleakconnection.py", line 62, in async_get_connection
raise BackendException("Can't find device")
custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.BackendException: Can't find device
2022-11-18 11:38:58.029 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Kacper] Broken connection [retry 7/10]: Can't find device
2022-11-18 11:38:59.039 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Kacper] Broken connection [retry 8/10]: Can't find device
2022-11-18 11:39:00.046 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Kacper] Broken connection [retry 9/10]: Can't find device
2022-11-18 11:39:01.053 WARNING (MainThread) [custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.bleakconnection] [Kacper] Broken connection [retry 10/10]: Can't find device
2022-11-18 11:39:01.060 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [547907073744] Can't find device
Traceback (most recent call last):
File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 200, in handle_call_service
await hass.services.async_call(
File "/usr/src/homeassistant/homeassistant/core.py", line 1744, in async_call
task.result()
File "/usr/src/homeassistant/homeassistant/core.py", line 1781, in _execute_service
await cast(Callable[[ServiceCall], Awaitable[None]], handler.job.target)(
File "/usr/src/homeassistant/homeassistant/helpers/entity_component.py", line 208, in handle_service
await service.entity_service_call(
File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 678, in entity_service_call
future.result() # pop exception if have
File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 943, in async_request_call
await coro
File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 715, in _handle_entity_call
await result
File "/config/custom_components/dbuezas_eq3btsmart/climate.py", line 245, in async_set_hvac_mode
await self._thermostat.async_set_mode(HA_TO_EQ_HVAC[hvac_mode])
File "/config/custom_components/dbuezas_eq3btsmart/python_eq3bt/eq3bt/eq3btsmart.py", line 315, in async_set_mode
return await self._async_set_mode(0)
File "/config/custom_components/dbuezas_eq3btsmart/python_eq3bt/eq3bt/eq3btsmart.py", line 340, in _async_set_mode
await self._conn.async_make_request(value)
File "/config/custom_components/dbuezas_eq3btsmart/python_eq3bt/eq3bt/bleakconnection.py", line 129, in async_make_request
raise ex
File "/config/custom_components/dbuezas_eq3btsmart/python_eq3bt/eq3bt/bleakconnection.py", line 111, in async_make_request
conn = await self.async_get_connection()
File "/config/custom_components/dbuezas_eq3btsmart/python_eq3bt/eq3bt/bleakconnection.py", line 62, in async_get_connection
raise BackendException("Can't find device")
custom_components.dbuezas_eq3btsmart.python_eq3bt.eq3bt.BackendException: Can't find device
```

**Comment by dbuezas** [comment]
Mmmm can you put the PI right next to one of them to make sure it is not a signal strength issue?

**Comment by lkempf** [comment]
I have the same issue (also on a PI) and I think I have managed to fix it. I'm currently trying to find out which of my modifications actually solved the issue and will create a PR as soon as I have done so.

**Comment by lkempf** [comment]
I think I have identified the underlying issue. The homeassistant bluetooth stack relies heavily on listening to advertisements to track which devices are still available. However, there seem to be cases where either no advertisements are received or they are lost somewhere before reaching HA. This will cause HA to assume that the device isn't here even if it is.

I see two solutions for this (and I will try to do both):
- Create an issue with HA about this to find out how their bluetooth backend could handle this case better.
- Add a fallback to this integration that bypasses the homeassitant bluetooth stack if the device isn't found.

**Comment by dbuezas** [comment]
That's very interesting!
There is a third option which I didn't consider because this was working fine for me as is: to try to connect via mac even if no advertisement has been received.
@ikempf, can you try the latest master (commit: "add connection fallback" [92211b033](https://github.com/dbuezas/eq3btsmart/tree/92211b033c03d19bb7c131fd494046d43ab54a4e))? Maybe

**Comment by dbuezas** [comment]
I don't want to rise hopes too much though: although the last error _"Can't find device"_ is due to the missing advertisements (as you discovered), in the log far above the issue was a missing characteristic _"Characteristic d0e8434d-cd29-0996-af41-6c90f4e0eb2a not found"_

**Comment by lkempf** [comment]
For me the only errors are "Can't find device" so I'm somewhat optimistic. However, your fix still results in "No device found for address ..." and the controller is never notified of the connection request. I'm investigating why.

**Comment by lkempf** [comment]
```
class HaBleakClientWrapper(BleakClient):
  """Wrap the BleakClient to ensure it does not shutdown our scanner.
  
  If an address is passed into BleakClient instead of a BLEDevice,
  bleak will quietly start a new scanner under the hood to resolve
  the address. This can cause a conflict with our scanner. We need
  to handle translating the address to the BLEDevice in this case
  to avoid the whole stack from getting stuck in an in progress state
  when an integration does this.
  """
```

Ok I don't think working around this issue is possible without changes in HA.

**Comment by dbuezas** [comment]
ok, it looks like it is a very bad idea to do this then. Did you make sure this doesn't happen to work for you anyway?

**Comment by lkempf** [comment]
Homeassitant intercepts the creation of BleakClient and redirects it to HaBleakClientWrapper

**Comment by dbuezas** [comment]
Did you report this to Home Assistant? I think they are probably not testing much on raspberry pi with connectable devices

**Comment by lkempf** [comment]
I reported it and it got closed as won't fix. So I've just ordered yet another Bluetooth dongle that's on the compatibility list. I can properly connect with one of the thermostats using the integrated Bluetooth of the Pi but the rest are out of reach. So I hope that with the new dongle it will actually work.

**Comment by paulokow** [comment]
@ikempf, do you know what change is needed in the HA code to make this working? Can you post the code here?

**Comment by lkempf** [comment]
As a really quick and dirty fix you can just place a `return None` on the first line of `_async_check_unavailable` in `homeassistant/components/bluetooth/manager.py`.

**Comment by lkempf** [comment]
@paulokow But looking at your logs more carefully I doubt that this will completely fix your issues. It might help a bit but you seem to be running into additional issues. Do you fulfill the software requirements listed at https://www.home-assistant.io/integrations/bluetooth/?

**Comment by paulokow** [comment]
@lkempf thanks. I just just use official HASS image on RPI3, so I don't assume anything wrong with the s/w stack.

**Comment by dbuezas** [comment]
Could it be that HA thinks the bluetooth adapter cannot connect to devices?
To test that, you could try fetching the BLEDevice while setting connectable to false
```python
        ble_device = bluetooth.async_ble_device_from_address(
            self._hass,
            self._mac,
            connectable=False # <--- this
        )
```

**Comment by dbuezas** [comment]
https://github.com/Heckie75/eQ-3-radiator-thermostat#3-pair-bluetooth

This cli implementation has an excellent readme, and says BT should be dis- and re-enabled via thermostat wheel

**Comment by paulokow** [comment]
OK, I solved my problem.
It turned out there might have been some "remainings" of the past connections from HA prior to 2022.7 stored in the adapter that conflicted with the new stack.

After going through `bluetoothctl` and `disconnect XX:XX:XX:XX:XX:XX` for all devices, all thermostats connected clean and are running OK now.

@dbuezas thanks for all the support anyway.
Leaving the issue open for you to decide if some special handling, like "reset" is possible to add to the driver.

**Comment by dbuezas** [comment]
Great to hear! Thanks for getting back with the solution.
